// peezy/modules/theme-engine.js
// setup

	var path = require("path"),
	 	fileHelper = require("peezy-file-helper"),
	 	_ = require("underscore");

	var ThemeEngine = {};
	
// initialize
	
	ThemeEngine.init = function(app_config, site_config) {
		
		this.siteBase = app_config.siteBase;
		this.themesBase = path.join(app_config.siteBase, app_config.folders.themes);
		this.folder = app_config.folders.themes;
	
		this.siteConfig = _.clone(site_config);

		this.loadThemes();
	
	}
	
// load pages
	
	ThemeEngine.loadThemes = function() {
	
		var themes = fileHelper.getFolders(this.themesBase);	

		// for each theme
		for (var i = 0; i < themes.length; i++) {
		
			// if required files dont exist, delete from theme list
			var theme_config = path.join(this.themesBase, themes[i], "theme.json");
			var theme_index = path.join(this.themesBase, themes[i], "index.ejs");
		
			if (!fileHelper.exists([theme_config, theme_index])) themes.splice(i, 1);
		
		}
	
		this.themes = themes;	
	
	}
	
// set theme

	ThemeEngine.setTheme = function(theme, query) {

		// theme override	
		if (query.theme && themeEngine.themes.indexOf(query.theme) > -1) theme = query.theme;
					
		// set vars
		this.theme = theme;
		this.themePath = "/" + this.folder + "/" + theme;
		this.themeBase = path.join(this.themesBase, theme);
		
		// load config
		this.themeConfig = require(path.join(this.themeBase, "theme.json"));

	}	

// set template

	ThemeEngine.setTemplate = function(template, query) {

		// set template
		if (query.inc && this.templateExists(theme, "inc/" + query.inc)) this.template = "inc/" + query.inc;
		else if (query.template && this.templateExists(theme, query.template)) this.template = query.template;
		else this.template = "index";

	}
	
// if template exists

	ThemeEngine.templateExists = function(theme, template) {
	
		var filepath = path.join(this.themesBase, theme, template + ".ejs");
		if (fileHelper.exists(filepath)) return true;
		else return false;
	
	}	
	
// build locals

	ThemeEngine.buildLocals = function(site_config, req_locals) {

		var locals = _.clone(site_config);
		
		// set theme locals
		locals.theme = this.theme;
		locals.themePath = this.themePath;
						
		// if child theme theme
		if (this.themeConfig.parent) {
			// set parent theme locals
			locals.parent = this.themeConfig.parent;
			locals.parentPath = paths.themes_rel + "/" + this.themeConfig.parent;
		}
		
		_.extend(locals, req_locals);
		
		this.locals = locals;
	
	}

module.exports = ThemeEngine;
